rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,131 @@
+# Integrate `cosign` with `kpack`
+
+## Problem
+
+The signing solution currently supported by `kpack` is
+[Notary v1](https://github.com/theupdateframework/notary). However, Notary
+presents itself with several different challenges, including:
+- It is unable to sign container image digests. The user must always specify a
+tag to be signed.
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures.
+  - Currently, there is no built-in command for relocation on `cosign`, but
+    copying container images and signatures across registries using other
+    container CLI tools such as `crane` and then verifying with `cosign` is
+    possible.
+  - There is also a proposal for
+    [implementing a `cosign cp`](https://github.com/sigstore/cosign/issues/303)
+    command that would affect this copy as a single operation.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Create a new configuration in the `Image` resource that allow users to
+  configure image signing with `cosign` as the signing mechanism, specifying
+  the secret reference to get the private key and passphrase. The user can also
+  specify signature annotations as part of configuration:
+```yaml
+cosign:
```

Comment:
Apart from having this on the image resource - my use case was to have this on the controller side. 

This was mainly to ensure that *every* image built by kpack is signed with the same cosign key that the cluster administrator controls. Currently this restricts the keys to be present in user namespace. Could we also add a similar field on the controller side?

Reply:
@matthewmcnew - can you please provide help here?  I haven't seen any mention of configuration for the controller itself anywhere in the `kpack` documentation.  This seems like an implementation detail.

Reply:
Cluster wide image signing is intriguing because it would allow you verify every image produced from a kpack install. However, It is also unprecedented within kpack. The current notary image signing functionality is configured per image. A cluster-wide image signing solution would be a challenge to implement because the build pods may exist in any namespace and would need a mechanism to mount the signing credentials across namespaces.

Reply:
@samj1912 If cosign support was configured per image as specified in this RFC would that still be useful for you? It seems that you could enforce that kpack images are configured with cosign?

Reply:
We could yes, but then the cosign secrets would be in each users namespace and they could use that to sign images generated elsewhere.

Reply:
Could you use RBAC to prevent users from reading the cosign secrets?

Reply:
RBAC doesn't allow denying access to certain resources. You will have to instead only allow certain secret object names they they can use which doesn't seem great UX as users couldn't create arbitrary secrets anymore. You would also need a webhook to default these cosign secrets for each image. The other issue would be updating or managing these cosign secrets in each user namespace and syncing them together, which again is a non trivial issue.

Reply:
Another alternative is to create a CRD which is a cluster resource that can store cosign secrets and managing that through RBAC so that only the controller or the pods created by it could use/access it. That might actually be a good way to solve this problem.

Reply:
I would also prefer if the cosign field in the spec was an array as cosign allows multiple private keys to sign the same image. Alternatively, this cluster resource for cosign could be referenced in the cluster builder which will ensure that certain cluster builders if used generate images signed by a particular key.

Reply:
Yeah, RBAC really only works if you prevent secret read.

Reply:
I mentioned this in the OSS CNB slack but, I fear a cluster wide signing solution may be outside the scope of kpack. However, it would definitely be possible with an extension controller that runs in cluster and automatically signs every image that is available on a build.status.latestImage.

Reply:
It sounds like, putting the `cosign` signing on the `Image` (feature 1) resource is helpful, but does not necessarily cover @samj1912 's entire use case (feature 2 - signing at the controller level).  I think we could approach both features separately.  Feature 1 is easier to implement and we could start sooner.  Feature 2 needs more design discussion before implementation could begin.  We could approach it after feature 1 is complete.  Agile programming for the win, right?

Additionally, it seems like a simple modification to make the `Image` resource take a list of `cosign` private keys and sign with each one.  We will make that change to the RFC.

Reply:
Just edited it to accept a list. I believe this should address the multiple key use case.

Reply:
@matthewmcnew, @dlorenc and I were discussing a possible implementation that doesn't require cross namespace secret mounts.

See this thread - https://sigstore.slack.com/archives/C01PZKDL4DP/p1623002688061600?thread_ts=1623002688.061600&cid=C01PZKDL4DP

The gist of it is -

A pod runs in user namespace and outputs the image ref and digest (so that the user doesn't have access to the signing key)
The controller can take just that information and generate the signature and upload it to a separate repository.

This way, no secrets from the user are needed or used.

Reply:
Maybe something to discuss in the WG meeting tomorrow?

Reply:
Sure thing!

Reply:
I believe this is being addressed on #765. Is it OK if I resolve this conversation?